sharding/nodesSetup.go

```diff
@@ -139,35 +168,69 @@ func (ns *NodesSetup) processShardAssignment() {
 	}
 }
 
-func (ns *NodesSetup) createInitialNodesPubKeys() {
+func (ns *NodesSetup) createInitialNodesInfo() {
 	nrOfShardAndMeta := ns.nrOfShards
 	if ns.MetaChainActive {
 		nrOfShardAndMeta += 1
 	}
 
-	ns.allNodesPubKeys = make(map[uint32][]string, nrOfShardAndMeta)
+	ns.allNodesInfo = make(map[uint32][]*NodeInfo, nrOfShardAndMeta)
 	for _, in := range ns.InitialNodes {
-		if in.pubKey != nil {
-			ns.allNodesPubKeys[in.assignedShard] = append(ns.allNodesPubKeys[in.assignedShard], string(in.pubKey))
+		if in.pubKey != nil && in.address != nil {
+			ns.allNodesInfo[in.assignedShard] = append(ns.allNodesInfo[in.assignedShard],
+				&NodeInfo{in.assignedShard, in.pubKey, in.address})
 		}
 	}
 }
 
-// InitialNodesPubKeys - gets initial public keys
+// InitialNodesPubKeys - gets initial nodes public keys
 func (ns *NodesSetup) InitialNodesPubKeys() map[uint32][]string {
-	return ns.allNodesPubKeys
+	allNodesPubKeys := make(map[uint32][]string, 0)
+	for shardId, nodesInfo := range ns.allNodesInfo {
+		pubKeys := make([]string, 0)
+		for i := 0; i < len(nodesInfo); i++ {
+			pubKeys = append(pubKeys, string(nodesInfo[i].pubKey))
+		}
+
+		allNodesPubKeys[shardId] = pubKeys
+	}
+
+	return allNodesPubKeys
+}
+
+// InitialNodesInfo - gets initial nodes info
+func (ns *NodesSetup) InitialNodesInfo() map[uint32][]*NodeInfo {
+	return ns.allNodesInfo
 }
 
-// InitialNodesPubKeysForShard - gets initial public keys
+// InitialNodesPubKeysForShard - gets initial nodes public keys for shard
 func (ns *NodesSetup) InitialNodesPubKeysForShard(shardId uint32) ([]string, error) {
-	if ns.allNodesPubKeys[shardId] == nil {
+	if ns.allNodesInfo[shardId] == nil {
+		return nil, ErrShardIdOutOfRange
+	}
+	if len(ns.allNodesInfo[shardId]) == 0 {
+		return nil, ErrNoPubKeys
+	}
+
+	nodesInfo := ns.allNodesInfo[shardId]
+	pubKeys := make([]string, 0)
+	for i := 0; i < len(nodesInfo); i++ {
+		pubKeys = append(pubKeys, string(nodesInfo[i].pubKey))
```

Comment:
same as above if you already know the size would be more efficient to allocate the right size and add on position instead of append.

Reply:
Done